packages/react-query-next-experimental/src/ReactQueryStreamedHydration.tsx

```diff
@@ -0,0 +1,87 @@
+'use client'
+
+import type { DehydratedState, QueryClient } from '@tanstack/react-query'
+import { dehydrate, hydrate, useQueryClient } from '@tanstack/react-query'
+import * as React from 'react'
+import type { HydrationStreamProviderProps } from './HydrationStreamProvider'
+import { createHydrationStreamProvider } from './HydrationStreamProvider'
+
+const stream = createHydrationStreamProvider<DehydratedState>()
+
+/**
+ * This component is responsible for:
+ * - hydrating the query client on the server
+ * - dehydrating the query client on the server
+ */
+export function ReactQueryStreamedHydration(props: {
+  children: React.ReactNode
+  queryClient?: QueryClient
+  transformer?: HydrationStreamProviderProps<DehydratedState>['transformer']
+}) {
+  const queryClient = useQueryClient(props.queryClient)
+
+  // <server only>
+  const isSubscribed = React.useRef(false)
+  /**
+   * We need to track which queries were added/updated during the render
+   */
+  const [trackedKeys] = React.useState(() => new Set<string>())
```

Comment:
Could this be a ref instead? Just thinking it would make it clearer that it's not normal state but being mutated directly.

Reply:
I think using state for this is actually the more ergonomic approach, we're also doing this in e.g. `useQuery`:

https://github.com/TanStack/query/blob/3643b87c34d86117b150ca845a058991ff132ab0/packages/react-query/src/useBaseQuery.ts#L68-L74

I also have a blogpost on that topic: https://tkdodo.eu/blog/use-state-for-one-time-initializations

Reply:
Oh yeah, I totally spaced actually and forgot refs don't support initializer functions.. 😬 Should also have said this was very much a NIT to start with.